src/ast/semantic_analyser.cpp

```diff
@@ -2820,6 +2818,83 @@ bool SemanticAnalyser::check_symbol(const Call &call, int arg_num __attribute__(
   return true;
 }
 
+bool SemanticAnalyser::check_available(const Call &call, ProbeType type)
+{
+  auto &func = call.func;
+
+  if (func == "reg")
+  {
+    switch (type)
+    {
+      case ProbeType::kprobe:
+      case ProbeType::kretprobe:
+      case ProbeType::uprobe:
+      case ProbeType::uretprobe:
+      case ProbeType::usdt:
+      case ProbeType::profile:
+      case ProbeType::interval:
+      case ProbeType::software:
+      case ProbeType::hardware:
+      case ProbeType::watchpoint:
+        return true;
+      case ProbeType::invalid:
+      case ProbeType::tracepoint:
+      case ProbeType::kfunc:
+      case ProbeType::kretfunc:
+        return false;
+    }
+  }
+  else if (func == "uaddr")
+  {
+    switch (type)
+    {
+      case ProbeType::usdt:
+      case ProbeType::uretprobe:
+      case ProbeType::uprobe:
+        return true;
+      case ProbeType::invalid:
+      case ProbeType::kprobe:
+      case ProbeType::kretprobe:
+      case ProbeType::tracepoint:
+      case ProbeType::profile:
+      case ProbeType::interval:
+      case ProbeType::software:
+      case ProbeType::hardware:
+      case ProbeType::watchpoint:
+      case ProbeType::kfunc:
+      case ProbeType::kretfunc:
+        return false;
+    }
+  }
+  else if (func == "signal")
+  {
+    switch (type)
+    {
+      case ProbeType::kprobe:
+      case ProbeType::kretprobe:
+      case ProbeType::uprobe:
+      case ProbeType::uretprobe:
+      case ProbeType::usdt:
+      case ProbeType::tracepoint:
+      case ProbeType::profile:
+      case ProbeType::kfunc:
+      case ProbeType::kretfunc:
+        return true;
+      case ProbeType::invalid:
+      case ProbeType::interval:
+      case ProbeType::software:
+      case ProbeType::hardware:
+      case ProbeType::watchpoint:
+        return false;
+    }
+  }
+
+  if (type == ProbeType::invalid)
```

Comment:
can we move this to the top and then remove all the invalid cases from the switch statements? Or does that trigger a missing case warning?

Reply:
yes, -Wswitch emits warnings. `warning: enumeration value 'invalid' not handled in switch [-Wswitch]`